simplefi-curve-exchange/src/curveUtil.ts

```diff
@@ -1,235 +1,299 @@
+  }
+  updateMarket(event, market, inputTokenBalances, pool.totalSupply);
 
-    return pool
+  return pool;
 }
 
 export function getPlainPoolInfo(pool: Address, coinCount: i32): PoolInfo {
-    let swapContract = StableSwapPlain3.bind(pool)
+  let swapContract = StableSwapPlain3.bind(pool);
 
-    let coins: Address[] = []
-    let balances: BigInt[] = []
+  let coins: Address[] = [];
+  let balances: BigInt[] = [];
 
-    let c: ethereum.CallResult<Address>
-    let b: ethereum.CallResult<BigInt>
+  let c: ethereum.CallResult<Address>;
+  let b: ethereum.CallResult<BigInt>;
 
-    for (let i = 0; i < coinCount; i++) {
-        let ib = BigInt.fromI32(i)
-        c = swapContract.try_coins(ib)
-        b = swapContract.try_balances(ib)
+  for (let i = 0; i < coinCount; i++) {
+    let ib = BigInt.fromI32(i);
+    c = swapContract.try_coins(ib);
+    b = swapContract.try_balances(ib);
 
-        if (!c.reverted && c.value.toHexString() != ADDRESS_ZERO && !b.reverted) {
-            coins.push(c.value)
-            balances.push(b.value)
-        }
+    if (!c.reverted && c.value.toHexString() != ADDRESS_ZERO && !b.reverted) {
+      coins.push(c.value);
+      balances.push(b.value);
     }
+  }
 
-    return {
-        coins,
-        underlyingCoins: [],
-        balances
-    }
+  return {
+    coins,
+    underlyingCoins: [],
+    balances,
+  };
 }
 
 export function getLendingPoolInfo(pool: Address, coinCount: i32): PoolInfo {
-    let swapContract = StableSwapLending3.bind(pool)
+  let coins: Address[] = [];
+  let underlyingCoins: Address[] = [];
+  let balances: BigInt[] = [];
 
-    let coins: Address[] = []
-    let underlyingCoins: Address[] = []
-    let balances: BigInt[] = []
+  let c: ethereum.CallResult<Address>;
+  let u: ethereum.CallResult<Address>;
+  let b: ethereum.CallResult<BigInt>;
 
-    let c: ethereum.CallResult<Address>
-    let u: ethereum.CallResult<Address>
-    let b: ethereum.CallResult<BigInt>
+  switch (coinCount) {
+    case 3:
+      let contract3 = StableSwapLending3.bind(pool);
 
-    for (let i = 0; i < coinCount; i++) {
-        let ib = BigInt.fromI32(i)
-        c = swapContract.try_coins(ib)
-        u = swapContract.try_underlying_coins(ib)
-        b = swapContract.try_balances(ib)
+      for (let i = 0; i < coinCount; i++) {
+        let ib = BigInt.fromI32(i);
+        c = contract3.try_coins(ib);
+        u = contract3.try_underlying_coins(ib);
+        b = contract3.try_balances(ib);
 
         if (!c.reverted && c.value.toHexString() != ADDRESS_ZERO && !b.reverted && !u.reverted) {
-            coins.push(c.value)
-            underlyingCoins.push(u.value)
-            balances.push(b.value)
+          coins.push(c.value);
+          underlyingCoins.push(u.value);
+          balances.push(b.value);
         }
-    }
+      }
+      break;
 
-    return {
-        coins,
-        underlyingCoins,
-        balances
-    }
+    case 4:
```

Comment:
It does not work for me with same ABI. Example of error I'm getting:

> graph-node_1  | Jul 21 18:05:11.593 ERRO Handler skipped due to execution failure, error: wasm trap: unreachable	wasm backtrace:	    0: 0x1157 - <unknown>!~lib/array/Array<~lib/@graphprotocol/graph-ts/chain/ethereum/ethereum.Value>#__get	    1: 0x2e51 - <unknown>!src/curveUtil/updatePool	    2: 0x2f9a - <unknown>!src/yPool/handleAddLiquidity	, handler: handleAddLiquidity, block_hash: 0x8e5efbcfe5bdd05836d15a62270a98c5dd81a382decf9798af46e7c1d42ea6b8, block_number: 9476528, sgd: 36, subgraph_id: QmUPzR3iHrz6nrb6BKStSUcp8bWHn99MwLxUepUXhvZcm4, component: SubgraphInstanceManager

I think the reason is that function signatures are not exactly the same, due to Curve changing contract implementation.  

I.e. to get balances Aave pool uses `balances(uint256):(uint256)` , but Y-pool uses `balances(int128):(uint256)`

Reply:
okay, Yea they mentioned in documentation that there are two version of Vyper, older version code uses `in128` for array index access and new one uses `uint256`  for array index access. I think we can restructure code to reflect that it's because of new or old Vyper instead of number of coins in the pool, that way we will have only two versions of the code instead of multiple version for each coin count.

Reply:
Updated, now `isUsingOldApi` parameter (defined in constants for each pool) is used to determine ABI to be used. I tried to hide all implementation details inside `getPoolInfo` and `getPoolBalances`.